File: 3771.total-score-of-dungeon-runs.py
#
# @lc app=leetcode id=3771 lang=python3
#
# [3771] Total Score of Dungeon Runs
#

# @lc code=start
import bisect
from typing import List

class Solution:
    def totalScore(self, hp: int, damage: List[int], requirement: List[int]) -> int:
        n = len(damage)
        prefix = [0] * (n + 1)
        for i in range(1, n + 1):
            prefix[i] = prefix[i - 1] + damage[i - 1]
        ans = 0
        for i in range(n):
            thresh = prefix[i + 1] + requirement[i] - hp
            left = bisect.bisect_left(prefix, thresh, 0, i + 1)
            ans += i + 1 - left
        return ans
# @lc code=end

File: 3771.total-score-of-dungeon-runs.java
# @lc app=leetcode id=3771 lang=java
#
# [3771] Total Score of Dungeon Runs
#
# @lc code=start
class Solution {
    public long totalScore(int hp, int[] damage, int[] requirement) {
        int n = damage.length;
        long[] prefix = new long[n + 1];
        for (int i = 0; i < n; ++i) {
            prefix[i + 1] = prefix[i] + damage[i];
        }
        long total = 0;
        for (int start = 0; start < n; ++start) {
            long curHp = hp;
            for (int i = start; i < n; ++i) {
                curHp -= damage[i];
                if (curHp >= requirement[i]) {
                    total++;
                }
            }
        }
        return total;
    }
}
# @lc code=end